### ISSUE TITLE
Add Support for Cognito User Pool Authentication

### ISSUE THREAD

**Issue by mipnw** [aws/aws-sdk-net#909]
**Feature Request**: add authentication for Cognito User Pool.

Currently the AWS SDK for .NET does not support Cognito User Pool Authentication. It does support Cognito User pool SignUp and ConfirmSignUp. But that is not sufficient. Once a user exists, they need to be able to authenticate.

This prevents C# applications (Xamarin, Unity, ASP.NET etc...) from being able to use Cognito User Pools which is a shame because that in turns prevents C# app development for the rest of the AWS stack  unless you can deal with unauthenticated workflows, or using a third party identity provider (AWS Cognito Identity Pool Federated to IdP) without using an AWS user pool.

This feature has been requested many times before, and yet it was closed each time, as a duplicate of another open issue, that was later closed as a duplicate of another until finally two were closed as duplicate of each other at which point the snake ate its tail. See
[`Issue #36`](https://github.com/awslabs/aws-sdk-xamarin/issues/36), [`Issue #390`](https://github.com/aws/aws-sdk-net/issues/390), [`Issue #495`](https://github.com/aws/aws-sdk-net/issues/495), [`Issue #590`](https://github.com/aws/aws-sdk-net/issues/590)

I am aware someone wrote [AWS SRP Login in C#](http://blog.mmlac.com/aws-cognito-srp-login-c-sharp-dot-net/). That solution references another library `Org.BouncyCastle`, has unclear licensing requirements, and is not production quality (the author documents the limitations).

**Comment by mipnw** [comment]
I found [AWSSDK.Extensions.CognitoAuthentication](https://www.nuget.org/packages/AWSSDK.Extensions.CognitoAuthentication). 

It seems to be a great library, I only found one bug in the last hour of testing various scenarios that interest me. It's too bad that it was very difficult to find this library when I searched for documentation on AWS Cognito User Pool this did not come up. It would be nice if this had been inside the AWS SDK for .NET rather than built as an extension, i.e. folded into NuGet package AWSSDK.CognitoIdentityProvider which is the package you find when you follow AWS tutorials
